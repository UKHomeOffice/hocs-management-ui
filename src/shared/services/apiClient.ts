import axios from 'axios';

export default {
    createClient: () => {
        const client = axios.create();

        return {
            get: client.get,
            delete: client.delete,
            post: client.post,
            put: client.put
        };
    }
};
