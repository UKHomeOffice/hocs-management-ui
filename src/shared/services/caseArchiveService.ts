import axios from 'axios';
import ArchiveCaseModel from '../models/archiveCaseModel';

export const archiveCase = (archiveCaseModel: ArchiveCaseModel) => new Promise((resolve, reject) => axios
    .post('/api/case/archive', archiveCaseModel)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
