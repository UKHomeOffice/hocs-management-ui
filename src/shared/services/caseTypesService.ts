import axios from 'axios';
import CaseType from '../models/caseType';
import Item from '../models/item';

const sortByLabel: ((a: Item, b: Item) => number) = (a, b) => {
    const labelA = a.label.toLocaleLowerCase();
    const labelB = b.label.toLocaleLowerCase();
    let comparison = 0;
    if (labelA > labelB) {
        comparison = 1;
    } else if (labelA < labelB) {
        comparison = -1;
    }
    return comparison;
};

export const getCaseTypes = () => new Promise<CaseType[]>((resolve, reject) =>
    axios.get('/api/case-types')
        .then((response) => {
            resolve(response.data.sort(sortByLabel));
        })
        .catch(error => reject(error))
);
