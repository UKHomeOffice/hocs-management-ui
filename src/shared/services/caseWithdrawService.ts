import axios from 'axios';
import WithdrawCaseModel from '../models/withdrawCaseModel';

export const withdrawCase = (withdrawCaseModel: WithdrawCaseModel) => new Promise((resolve, reject) => axios
    .post('/api/case/withdraw', withdrawCaseModel)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
