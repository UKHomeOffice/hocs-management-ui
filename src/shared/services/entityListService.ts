import axios from 'axios';
import EntityListItem from 'shared/models/entityListItem';

export const getListItems = (listName: string) => new Promise<EntityListItem[]>((resolve, reject) => axios
    .get(`/api/entity/list/${listName}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const createListItem = (item: EntityListItem, listName: string) => new Promise((resolve, reject) => axios
    .post(`/api/entity/list/${listName}`, item)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);

export const getItemDetails = (itemUUID: string) => new Promise<EntityListItem>((resolve, reject) => axios
    .get(`/api/entity/${itemUUID}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const updateListItem = (item: EntityListItem, listName: string) => new Promise((resolve, reject) => axios
    .post(`/api/entity/list/update/${listName}`, item)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);

export const deleteListItem = (uuid: string, listName: string) => new Promise((resolve, reject) => axios
    .delete(`/api/entity/list/${listName}`, { data: { entityUUID: uuid } })
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
