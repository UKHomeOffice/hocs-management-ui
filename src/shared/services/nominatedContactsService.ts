import axios from 'axios';
import NominatedContact from "../models/nominatedContact";

export const addNominatedContact = (nominatedContact: NominatedContact) => new Promise((resolve, reject) => axios
    .post('/api/nominated-contact', nominatedContact)
    .then(() => resolve())
    .catch(reason => reject(reason))
);
