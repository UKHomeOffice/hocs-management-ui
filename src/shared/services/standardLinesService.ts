import axios from 'axios';
import UpdateRequest from 'shared/pages/standardLineAmend/updateRequest';

export const addStandardLine = (standardLine: FormData) => new Promise((resolve, reject) => axios
    .post('/api/standard-lines', standardLine)
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const updateStandardLine = (standardLine: UpdateRequest) => new Promise((resolve, reject) => axios
    .post('/api/standard-lines/update', standardLine)
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const getAllStandardLines = () => new Promise((resolve, reject) => axios
    .get('/api/standard-lines/all')
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const getStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .get(`/api/standard-lines/get/${standardLineUuid}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const expireStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .post(`/api/standard-lines/expire/${standardLineUuid}`, {})
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const deleteStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .delete(`/api/standard-lines/delete/${standardLineUuid}`, {})
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const downloadStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .get(`/api/standard-lines/download/${standardLineUuid}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);
