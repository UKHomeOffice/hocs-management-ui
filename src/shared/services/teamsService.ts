import axios from 'axios';
import Item from '../models/item';
import Team from '../models/team';

export const getTeam = (teamId: string) => new Promise<Team>((resolve, reject) =>
    axios.get(`/api/teams/${teamId}`)
        .then(response => resolve(response.data))
        .catch(error => reject(error))
);

export const getTeams = () => new Promise<[Item]>((resolve, reject) =>
    axios.get('/api/teams')
        .then(response => resolve(response.data))
        .catch(error => reject(error))
);

export const getTeamMembers = (teamId: string) => new Promise((resolve, reject) => axios
    .get(`/api/teams/${teamId}/members`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const getTeamsForUser = (userId: string) => new Promise<[Item]>((resolve, reject) => axios
    .get(`/api/teams/${userId}/teams`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const createTeam = (team: Team) => new Promise((resolve, reject) => axios
    .post(`/api/teams/unit/${team.unitUUID}`, team)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);
