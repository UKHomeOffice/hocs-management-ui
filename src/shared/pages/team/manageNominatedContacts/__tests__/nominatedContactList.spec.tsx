import { act, fireEvent, getByText, render, RenderResult, wait } from '@testing-library/react';
import React from 'react';
import NominatedContactList from '../nominatedContactList';
import Contact from '../../../../models/contact';
import Item from '../../../../models/item';
import * as NominatedContactsService from '../../../../services/nominatedContactsService';

const mockContacts: Array<Contact> = [
    { id: 0, uuid: 'uuid1', teamUUID: 'teamUUID1', emailAddress: 'one@example.org' },
    { id: 1, uuid: 'uuid2', teamUUID: 'teamUUID2', emailAddress: 'two@example.org' },
    { id: 2, uuid: 'uuid3', teamUUID: 'teamUUID3', emailAddress: 'three@example.org' }
];

const mockContactShortList: Array<Contact> = [
    { id: 0, uuid: 'uuid1', teamUUID: 'teamUUID1', emailAddress: 'one@example.org' }
];

const mockContactItems: Array<Item> = [
    { value: 'uuid1', label: 'one@example.org' },
    { value: 'uuid2', label: 'two@example.org' },
    { value: 'uuid3', label: 'three@example.org' }
];

const mockContactItemsShortList: Array<Item> = [
    { value: 'uuid1', label: 'one@example.org' }
];

const getNominatedContactsForTeamSpy =
    jest.spyOn(NominatedContactsService, 'getNominatedContactsForTeam');

const removeNominatedContactsForTeamSpy =
    jest.spyOn(NominatedContactsService, 'removeNominatedContactFromTeam');

const mockClearErrors = jest.fn();
const mockSetErrorMessage = jest.fn();
const mockDispatch = jest.fn();

const renderComponent = () => render(
    <NominatedContactList teamId="testTeam" errorFuncs={[null, null, mockClearErrors, mockSetErrorMessage]}/>
);

describe('NominatedContactList', () => {
    beforeEach(() => {
        jest.spyOn(React, 'useContext').mockImplementation(() => ({
            dispatch: mockDispatch,
            state: {
                contacts: mockContactItems
            }
        }));

        mockClearErrors.mockReset();
        mockSetErrorMessage.mockReset();
        mockDispatch.mockReset();
    });

    describe('when the NominatedContactList component is mounted', () => {
        it('should render with default props and populate state', async () => {
            getNominatedContactsForTeamSpy.mockImplementationOnce(() => Promise.resolve(mockContacts));

            let wrapper: RenderResult;
            act(() => {
                wrapper = renderComponent();
            });
            expect.assertions(2);

            await wait(() => {
                expect(mockDispatch).toBeCalledWith({ 'type': 'SetContacts', payload: mockContactItems });
                expect(wrapper.container).toMatchSnapshot();
            });
        });

        it('should set an error if get nominated contacts failed', async () => {
            getNominatedContactsForTeamSpy.mockImplementationOnce(() => Promise.reject({
                response: {
                    status: 500
                }
            }));

            let wrapper: RenderResult;
            act(() => {
                wrapper = renderComponent();
            });
            expect.assertions(3);

            await wait(() => {
                expect(mockDispatch).toBeCalledTimes(0);
                expect(mockSetErrorMessage).toBeCalledWith({
                    'description': 'There was an error retrieving the list of nominated contacts.  Please try refreshing the page.',
                    'title': 'Something went wrong'
                });

                expect(wrapper.container).toMatchSnapshot();
            });
        });
    });

    describe('remove contact', () => {
        let wrapper: RenderResult;
        beforeEach(() => {
            getNominatedContactsForTeamSpy.mockImplementationOnce(() => Promise.resolve(mockContacts));
            jest.spyOn(React, 'useContext').mockImplementation(() => ({
                dispatch: mockDispatch,
                state: {
                    contacts: mockContactItems
                }
            }));

            act(() => {
                wrapper = renderComponent();
            });
        });

        it('should remove a contact when remove is clicked', async () => {
            removeNominatedContactsForTeamSpy.mockImplementationOnce(() => Promise.resolve());

            act(() => {
                const selectedContact = getByText(wrapper.container, 'two@example.org');
                const row = (selectedContact.closest('tr'));
                const removeButton = getByText(row as HTMLElement, 'Remove');
                fireEvent.click(removeButton);
            });

            await wait(() => {
                expect(mockDispatch).toBeCalledTimes(2);
                expect(mockDispatch).toBeCalledWith({
                    'type': 'RemoveContact',
                    'payload': { 'value': 'uuid2', 'label': 'two@example.org' }
                });
            });
        });

        it('should set an error if the item cannot be removed', async () => {
            removeNominatedContactsForTeamSpy.mockImplementationOnce(() => Promise.reject({
                response: {
                    status: 500
                }
            }));

            act(() => {
                const selectedContact = getByText(wrapper.container, 'two@example.org');
                const row = (selectedContact.closest('tr'));
                const removeButton = getByText(row as HTMLElement, 'Remove');
                fireEvent.click(removeButton);
            });

            await wait(() => {
                expect(mockDispatch).toBeCalledTimes(1);
                expect(mockSetErrorMessage).toBeCalledWith({
                    'description': 'There was an error deleting the nominated contact. Please try refreshing the page.',
                    'title': 'Something went wrong',
                });
            });
        });
    });

    describe('Remove contact messaging', () => {
        beforeEach(() => {
            jest.spyOn(React, 'useContext').mockImplementation(() => ({
                dispatch: mockDispatch,
                state: {
                    contacts: mockContactItemsShortList
                }
            }));

            mockClearErrors.mockReset();
            mockSetErrorMessage.mockReset();
            mockDispatch.mockReset();
        });

        it('should set an error if there is only one contact', async () => {
            getNominatedContactsForTeamSpy.mockImplementationOnce(() => Promise.resolve(mockContactShortList));

            let wrapper: RenderResult;
            act(() => {
                wrapper = renderComponent();
                const selectedContact = getByText(wrapper.container, 'one@example.org');
                const row = (selectedContact.closest('tr'));
                const removeButton = getByText(row as HTMLElement, 'Remove');
                fireEvent.click(removeButton);
            });

            await wait(() => {
                expect(mockDispatch).toBeCalledTimes(1);
                expect(mockSetErrorMessage).toBeCalledWith({
                    'description': 'Unable to delete nominated contact - teams must have at least one contact.',
                    'title': 'Something went wrong',
                });
            });
        });
    });
});
