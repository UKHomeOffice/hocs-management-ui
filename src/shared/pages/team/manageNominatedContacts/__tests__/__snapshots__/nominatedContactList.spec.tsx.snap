// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NominatedContactList when the NominatedContactList component is mounted should render with default props and populate state 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <table
        class="govuk-table"
      >
        <caption
          class="govuk-table__caption"
        >
          Nominated contacts:
        </caption>
        <tbody
          class="govuk-table__body"
        >
          <tr
            class="govuk-table__row"
          >
            <th
              class="govuk-table__header"
              scope="row"
            >
              one@example.org
            </th>
            <td
              class="govuk-table__cell"
            >
              <a
                class="govuk-link"
                href="#"
              >
                Remove
                <span
                  class="govuk-visually-hidden"
                >
                   contact
                </span>
              </a>
            </td>
          </tr>
          <tr
            class="govuk-table__row"
          >
            <th
              class="govuk-table__header"
              scope="row"
            >
              two@example.org
            </th>
            <td
              class="govuk-table__cell"
            >
              <a
                class="govuk-link"
                href="#"
              >
                Remove
                <span
                  class="govuk-visually-hidden"
                >
                   contact
                </span>
              </a>
            </td>
          </tr>
          <tr
            class="govuk-table__row"
          >
            <th
              class="govuk-table__header"
              scope="row"
            >
              three@example.org
            </th>
            <td
              class="govuk-table__cell"
            >
              <a
                class="govuk-link"
                href="#"
              >
                Remove
                <span
                  class="govuk-visually-hidden"
                >
                   contact
                </span>
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`NominatedContactList when the NominatedContactList component is mounted should set an error if get nominated contacts failed 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <table
        class="govuk-table"
      >
        <caption
          class="govuk-table__caption"
        >
          Nominated contacts:
        </caption>
        <tbody
          class="govuk-table__body"
        >
          <tr
            class="govuk-table__row"
          >
            <th
              class="govuk-table__header"
              scope="row"
            >
              one@example.org
            </th>
            <td
              class="govuk-table__cell"
            >
              <a
                class="govuk-link"
                href="#"
              >
                Remove
                <span
                  class="govuk-visually-hidden"
                >
                   contact
                </span>
              </a>
            </td>
          </tr>
          <tr
            class="govuk-table__row"
          >
            <th
              class="govuk-table__header"
              scope="row"
            >
              two@example.org
            </th>
            <td
              class="govuk-table__cell"
            >
              <a
                class="govuk-link"
                href="#"
              >
                Remove
                <span
                  class="govuk-visually-hidden"
                >
                   contact
                </span>
              </a>
            </td>
          </tr>
          <tr
            class="govuk-table__row"
          >
            <th
              class="govuk-table__header"
              scope="row"
            >
              three@example.org
            </th>
            <td
              class="govuk-table__cell"
            >
              <a
                class="govuk-link"
                href="#"
              >
                Remove
                <span
                  class="govuk-visually-hidden"
                >
                   contact
                </span>
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
